src/gui/settingsdialog.cpp

```diff
@@ -178,13 +189,15 @@ void SettingsDialog::reject()
     ConfigFile cfg;
     cfg.saveGeometry(this);
     QDialog::reject();
+    CocoaProcessTypeToBackground();
 }
 
 void SettingsDialog::accept()
 {
     ConfigFile cfg;
     cfg.saveGeometry(this);
     QDialog::accept();
+    CocoaProcessTypeToBackground();
```

Comment:
Regarding the Cocoa calls, I thought I mentioned this, but the interface I created could be generalized to work on other platforms, as well, in particular with the [QWinTaskbarButton Class](https://doc.qt.io/qt-5/qwintaskbarbutton.html). (I wasn't able to find any corresponding classes for the Linux dock.)

As far using platform-specific code goes... the fact of the matter is that Qt should be providing the `ToForeground()` and `ToBackground()` functionality, but it isn't, and sometimes it's necessary to provide workarounds for deficiencies in upstream projects.

Reply:
> As far using platform-specific code goes... the fact of the matter is that Qt should be providing the `ToForeground()` and `ToBackground()` functionality,

I'll respectfully (but strongly) disagree with this. ;-)

It *could* call it internally, but present an API for it. Nope.

> but it isn't, and sometimes it's necessary to provide workarounds for deficiencies in upstream projects.

I'm not disagreeing with this. I'm just disagreeing on the approach you chose to do so in the present case. An approach where it means adding more such calls to something which is platform specific like this in every dialog for a cross-platform project isn't viable (I'd claim the same if those were calls to QWinTaskbarButton by the way).

Luckily going through the event filter route wouldn't suffer from the same coupling issues and is way more viable.

Reply:
> adding more such calls to something which is platform specific like this in every dialog

The way an interface works is that you include the parent interface in the file where you're using the functions, then choose which of the implementing classes to import at compile time. (The interface is also necessary in order to prevent the C++ file from trying to include AppKit directly.) So, for example, the code could be restructured as:

* `foregroundbackground_interface.h`
* `foregroundbackground_linux.cpp`
* `foregroundbackground_mac.mm` (+ `foregroundbackground_cocoa.mm`)
* `foregroundbackground_windows.cpp`

We know that the Mac code would use `CocoaProcessType`, and the Windows code would use `QWinTaskbarButton`, but, again, it isn't clear to me what the Linux code would use, so it would remain a dummy class.

As for invocation, if it would work better to use events, then by all means, *you* can do so. I just don't know what would be involved. Indeed, this pull request is set up so that project maintainers can edit it...

Reply:
Well, could you at least point me in the direction of a Qt API reference page? I could mark this as a draft again and try to figure it out myself, since I kind of have all the time in the world right now, lol.

Reply:
> Well, could you at least point me in the direction of a Qt API reference page? I could mark this as a draft again and try to figure it out myself, since I kind of have all the time in the world right now, lol.

Well I'd expect the link I provided to cover it nicely, they give a lot of details to it:
https://doc.qt.io/qt-5/qobject.html#installEventFilter

Now the thing is that you would probably want to call installEventFilter with your own filter on the application object (the one in main.cpp) this is the object which should "see all". So the show/hide events of dialogs would go through it.

Reply:
@elsiehupp I like the idea of interface files and implementations for different platforms you mentioned:

- foregroundbackground_interface.h
- foregroundbackground_linux.cpp
- foregroundbackground_mac.mm (+ foregroundbackground_cocoa.mm)
- foregroundbackground_windows.cpp

Also, I agree with @er-vin and I don't like having **Cocoa**something in the code that's visible on all platforms. I'd rather keep such naming within the **foregroundbackground_mac.mm** and its included files.
Then, the interface would just have functions like:
- processToBackground()
- processToForeground();

Also, I'd put these functions into a namespace. Probably under:
**Utility::Something**? (Something could be a System or whatever makes sense).

Then, just call the functions with full namespace qualifiers:

- Utility::Something::processToBackground()
- Utility::Something::processToForeground();

As for another idea of using the event filter as @er-vin suggested - I fully support this too,

Reply:
I have a more recent commit that I forgot to push...